File: src/player/main/sp/it/pl/gui/objects/image/cover/ImageCover.java
package sp.it.pl.gui.objects.image.cover;

import java.awt.image.BufferedImage;
import java.io.File;
import java.util.Objects;
import javafx.embed.swing.SwingFXUtils;
import javafx.scene.image.Image;
import static sp.it.util.dev.FailKt.noNull;

/**
 * Denotes Cover represented by a {@link javafx.scene.image.Image} or {@link java.awt.image.BufferedImage}.
 * <p/>
 * This class is fully polymorphic
 * Should never be used directly but instead use the {@link sp.it.pl.gui.objects.image.cover.Cover} interface
 * and leverage polymorphism.
 */
public class ImageCover implements Cover {
	private final Image imageI;
	private final BufferedImage imageB;
	private final String info;

	public ImageCover(Image image, String description) {
		noNull(description);

		this.imageI = image;
		this.imageB = null;
		this.info = description;
	}

	public ImageCover(BufferedImage image, String description) {
		noNull(description);

		this.imageB = image;
		this.imageI = null;
		this.info = description;
	}

	@Override
	public Image getImage() {
		return imageB==null ? imageI : SwingFXUtils.toFXImage(imageB, null);
	}

	@Override
	public Image getImage(double width, double height) {
		return getImage();
	}

	@Override
	public File getFile() {
		return null;
	}

	@Override
	public boolean isEmpty() {
		return imageB==null && imageI==null;
	}

	@Override
	public String getDescription() {
		return info;
	}

	@Override
	public boolean equals(Object o) {
		if (o==this) return true;

		if (o!=null && o instanceof ImageCover) {
			ImageCover other = (ImageCover) o;
			if (imageB!=null && other.imageB!=null)
				return imageB.equals(other.imageB);
			if (imageI!=null && other.imageI!=null)
				return imageI.equals(other.imageI);
		}
		return false;
	}

	@Override
	public int hashCode() {
		int hash = 3;
		hash = 37*hash + Objects.hashCode(this.imageI);
		hash = 37*hash + Objects.hashCode(this.imageB);
		return hash;
	}

}

File: src/player/main/sp/it/pl/gui/objects/image/cover/ImageCover.kt
package sp.it.pl.gui.objects.image.cover

import javafx.embed.swing.SwingFXUtils
import javafx.scene.image.Image
import java.awt.image.BufferedImage
import java.util.Objects

/**
 * Denotes Cover represented by a [javafx.scene.image.Image] or [java.awt.image.BufferedImage].
 */
class ImageCover: Cover {
   private val imageI: Image?
   private val imageB: BufferedImage?
   private val info: String

   constructor(image: Image?, description: String) {
      this.imageI = image
      this.imageB = null
      this.info = description
   }

   constructor(image: BufferedImage?, description: String) {
      this.imageB = image
      this.imageI = null
      this.info = description
   }

   override fun getImage(): Image? = if (imageB==null) imageI else SwingFXUtils.toFXImage(imageB, null)

   override fun getImage(width: Double, height: Double): Image? = image

   override fun getFile(): Nothing? = null

   override fun isEmpty() = imageB==null && imageI==null

   override fun getDescription() = info

   override fun equals(other: Any?): Boolean {
      if (other===this) return true
      return other is ImageCover && info==other.info && imageI==other.imageI && bufferedImagesEqual(imageB, other.imageB)
   }

   override fun hashCode(): Int {
      var hash = 3
      hash = 37*hash + Objects.hashCode(this.imageI)
      hash = 37*hash + Objects.hashCode(this.imageB)
      return hash
   }

   companion object {
      fun bufferedImagesEqual(img1: BufferedImage?, img2: BufferedImage?): Boolean = when {
            img1===img2 -> true
            img1==img2 -> true
            img1==null || img2==null -> false
            img1.width!=img2.width || img1.height!=img2.height -> false
            else -> {
               fun e(): Boolean {
                  for (x in 0 until img1.width) {
                     for (y in 0 until img1.height) {
                        if (img1.getRGB(x, y)!=img2.getRGB(x, y))
                           return false
                     }
                  }
                  return true
               }
               e()
            }
         }
   }
}